pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java

```diff
@@ -295,4 +301,69 @@ public void globalLibrarySuccess() throws Exception {
         j.assertLogContains("running inside closure2", b);
 
     }
+
+    @Test
+    public void syntheticStages() throws Exception {
+        prepRepoWithJenkinsfile("syntheticStages");
+
+        WorkflowRun b = getAndStartBuild();
+        j.assertBuildStatusSuccess(j.waitForCompletion(b));
+
+        j.assertLogContains("[Pipeline] { (Tool Install)", b);
+        j.assertLogContains("[Pipeline] { (Checkout SCM)", b);
+        j.assertLogContains("[Pipeline] { (foo)", b);
+        j.assertLogContains("hello", b);
+        j.assertLogContains("[Pipeline] { (Post Build Actions)", b);
+        j.assertLogContains("I AM A POST-BUILD", b);
+
+        FlowExecution execution = b.getExecution();
+
+        Collection<FlowNode> heads = execution.getCurrentHeads();
+
+        DepthFirstScanner scanner = new DepthFirstScanner();
+
+        assertNotNull(scanner.findFirstMatch(heads, null, syntheticStagePredicate(SyntheticStage.toolInstall(), SyntheticStage.SYNTHETIC_PRE)));
+        assertNotNull(scanner.findFirstMatch(heads, null, syntheticStagePredicate(SyntheticStage.checkout(), SyntheticStage.SYNTHETIC_PRE)));
+        assertNotNull(scanner.findFirstMatch(heads, null, syntheticStagePredicate(SyntheticStage.postBuild(), SyntheticStage.SYNTHETIC_POST)));
+        assertNull(scanner.findFirstMatch(heads, null, syntheticStagePredicate(SyntheticStage.agentSetup(), SyntheticStage.SYNTHETIC_PRE)));
+    }
+
+    @Test
+    public void noToolSyntheticStage() throws Exception {
+        prepRepoWithJenkinsfile("noToolSyntheticStage");
+
+        WorkflowRun b = getAndStartBuild();
+        j.assertBuildStatusSuccess(j.waitForCompletion(b));
+
+        j.assertLogContains("[Pipeline] { (Checkout SCM)", b);
+        j.assertLogContains("[Pipeline] { (foo)", b);
+        j.assertLogContains("hello", b);
+        j.assertLogContains("[Pipeline] { (Post Build Actions)", b);
+        j.assertLogContains("I AM A POST-BUILD", b);
+
+        FlowExecution execution = b.getExecution();
+
+        Collection<FlowNode> heads = execution.getCurrentHeads();
+
+        DepthFirstScanner scanner = new DepthFirstScanner();
```

Comment:
Fun fact: this doesn't work with `LinearBlockHoppingScanner` - just tried it. We're post-build at this point, so as is said in that class's Javadoc, "ONLY use this with nodes inside the flow graph". =)